### ISSUE TITLE
Using deployment keys for private repositories with deis builder

### ISSUE THREAD

**Issue by mvgijssel** [deis/deis#2957]
Is it currently possible for the deis builder to access private repositories hosted on services like github / bitbucket using SSH?

**Comment by bacongobbler** [comment]
currently no. We don't have a way to attach docker auth credentials for a build.

**Comment by bacongobbler** [comment]
Relevent: we now have support for private buildpacks through #3259 

ping @iancoffey as I believe this is something we discussed :)

**Comment by iancoffey** [comment]
awesome! :+1:

**Comment by mboersma** [comment]
Now that #3305 is merged, the original issue might be fixed by using `deis config:set SSH_KEY=`.

**Comment by bacongobbler** [comment]
indeed, but I believe this issue is around telling the builder to pull a private repository from a git URL, rather than cloning a private repository locally then pushing it to the builder. We don't have support for the former, but we do for the latter!

Note that this is [how Engine Yard deploys apps](https://support.cloud.engineyard.com/hc/en-us/articles/205407628-Set-Up-Your-Application-and-Environment).

**Comment by mboersma** [comment]
Actually, #3305 implemented setting `BUILDPACK_URL` to a private GitHub repository specifically.

**Comment by mboersma** [comment]
@mvgijssel what was your requirement here specifically? Deis now handles the case where a private buildpack needs an SSH key for access, but we weren't sure if you implied larger changes.

**Comment by Crispy1975** [comment]
We have this issue also. Using the official Heroku Node.js buildpacks we need to pull in various GitHub private repos (modules) to build our app. Dokku gets around this with plugs to share deployment and host keys. However to date I have seen no way to build with private dependencies in our package.json file(s).

```npm ERR! Host key verification failed.
npm ERR! fatal: Could not read from remote repository.
npm ERR!
npm ERR! Please make sure you have the correct access rights
npm ERR! and the repository exists.```

**Comment by bacongobbler** [comment]
@Crispy1975 this seems related to https://github.com/heroku/heroku-buildpack-nodejs/issues/208, so yes it sounds like a buildpack-specific issue.

**Comment by Crispy1975** [comment]
@bacongobbler going to try their solution locally to see if that works. Will report back here.

**Comment by Crispy1975** [comment]
Nope, no luck. Forced 2.6.x of npm in package.json also logged in to the deis-builder and make sure it had known_hosts and relevant keys for GitHub access (assuming that is correct) and still getting the same error.

**Comment by bacongobbler** [comment]
@Crispy1975 would you mind opening a separate issue so we can track this independently? This sounds like a separate issue than the OP. Thanks!

**Comment by Crispy1975** [comment]
Was about to file a new issue and I noticed that I actually got a different error this time...

       $ git push -f deis testing/deis:master
       Counting objects: 424, done.
       Delta compression using up to 8 threads.
       Compressing objects: 100% (197/197), done.
       Writing objects: 100% (424/424), 45.00 KiB | 0 bytes/s, done.
       Total 424 (delta 249), reused 366 (delta 218)
       -----> Node.js app detected
              
       -----> Reading application state
              package.json...
              build directory...
              cache directory...
              environment variables...
              
              Node engine:         0.12.x
              Npm engine:          2.6.x
              Start mechanism:     npm start
              node_modules source: package.json
              node_modules cached: false
              
              NPM_CONFIG_PRODUCTION=true
              NODE_MODULES_CACHE=true
              
       -----> Installing binaries
              Resolving node version 0.12.x via semver.io...
              Downloading and installing node 0.12.5...
              Resolving npm version 2.6.x via semver.io...
              Downloading and installing npm 2.6.1 (replacing version 2.11.2)...


              ...


       npm ERR! ssh: connect to host github.com port 22: No route to host
       npm ERR! fatal: Could not read from remote repository.
       npm ERR!
       npm ERR! Please make sure you have the correct access rights
       npm ERR! and the repository exists.

ssh having issues getting through to GitHub on port 22. Yet inside the builder I had connectivity.

**Comment by Crispy1975** [comment]
Have started a new issue on the above here #3923

**Comment by bacongobbler** [comment]
ping @mvgijssel. Quoting @mboersma:

> What was your requirement here specifically? Deis now handles the case where a private buildpack needs an SSH key for access, but we weren't sure if you implied larger changes.

**Comment by mvgijssel** [comment]
Sorry for the late reply! Besides private buildpacks I was also interested in pulling private repositories for dependencies (gems, js modules, pip packages, ...). One of these issues is now fixed and the other is referenced in #3923.
